### DISCUSSION TITLE
More control over MTEXT rendering

### DISCUSSION THREAD

**Discussion by mbway** [mozman/ezdxf#623] [Ideas]
Currently either `draw_complex_mtext` or `draw_simple_mtext` can be called.

Complex
- renders text word by word
- renders all decorations, stretched words, lines (underline or fraction), filled polygons (backgrounds) etc
- does not support caret encoded newlines (although this is probably a bug)
- supports columns

Simple
- renders text line by line
- renders no decorations
- does not support caret encoded newlines or other layout commands
- does not support columns

Would it be possible to configure the mtext rendering so decorations for example are ignored but caret encoding and columns are still supported?

Also why is word-by-word rendering used? Is it because dxf only allows changing styles at word boundaries?
Would it be possible to combine adjacent 'cells' into a single cell for the purposes of rendering.

For example if a line of words are all drawn with the same style can't they be rendered as a single line rather than placing each word individually with it's own transform? This may provide a slight performance improvement for rendering images but it also would help when using the drawing addon for format conversion as it will be easier to determine the 'plain text' than if each word was placed independently.

**Comment by mozman** [comment]
The "word by word" question is easy to answer: by doing it "letter by letter" you have to handle kerning and ligatures by yourself - that's beyond my abilities and the APIs exposed by `Matplotlib`

**Reply by mbway** [reply]
but wouldn't it make more sense to render in chunks of 'runs of the same style' so if I did something like

'thisIsMy**TextWith**Style' you would render `thisIsMy` then `TextWith` then `Style`

and if you had

'this is my **text with** style' you would render `this is my ` then `text with` then ` style`

I was wondering whether word boundaries had any special significance other than a convenient place to break into chunks

**Reply by mozman** [reply]
The current implementation handles most of the complex MTEXT features:
![image](https://user-images.githubusercontent.com/4230966/150556232-7f6e83fb-3822-4c72-aec7-1b2199afe6e5.png)

There are also fractions, left-, right- and justified paragraph alignments, vertical alignments (top, middle center), tabulators,...

This is a single MTEXT entity:
![image](https://user-images.githubusercontent.com/4230966/150557130-0e313e73-3448-4488-af5d-7de75581459b.png)

I can't do better!

**Reply by mbway** [reply]
my issue is not with accuracy or lack of feature support. The above is very impressive! I was thinking about whether the internal representation could be improved, especially for chunks of text where complex formatting commands are not present.

**Reply by mozman** [reply]
I don't think combining adjacent cells is possible with the current implementation:

- the content-cells are created at first, the final properties for each cell like color, font or text size is already known
- the content-cell stores the width and height (calculated by Matplotlib) and the render properties, the final location is set by the layout engine, which then calls the render method of the cell
- the "gluing" can only be done after placing all cells by the layout engine
- to glue cells, the exact properties of the adjacent cell has to be known - only cells with exact the same properties can be joined together
- the layout engine knows nothing about the cell content, only the width, height and the location (after the layout calculation)
- the rendering process calls the render method of the cell with the target coordinates as arguments - therefore the layout engine cannot know if cells can be glued together
- the rendering method of the cell knows nothing about its surrounding
- the glue-cells between content-cells are not render cells, they are only spacers

I am not sure if this would speed up rendering, there is so much work to do in advance, the final rendering process itself is just a small part.
